src/plugins/vis_augmenter/public/vega/helpers.ts

```diff
@@ -0,0 +1,328 @@
+
+/**
+ * By default, the source datatable will not include rows with empty data.
+ * For handling events that may belong in missing buckets that are not yet
+ * created, we need to create them. For more details, see description in
+ * https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3145
+ *
+ * Note that this may add buckets with start/end times out of the chart bounds.
+ * This is the current default behavior of histogram aggregations with intervals,
+ * in order for the bucket keys to have "clean" timestamp keys (e.g., 1/1 @ 12AM).
+ * For more details, see
+ * https://opensearch.org/docs/latest/opensearch/bucket-agg/#histogram-date_histogram
+ *
+ * Also note this is only adding empty buckets at the beginning/end of a table. We are
+ * not taking into account missing buckets within source datapoints. Because of this
+ * limitation, it is possible that charted events may not be put into the most precise
+ * bucket based on their raw event timestamps, if there is missing / sparse source data.
+ */
+export const addMissingRowsToTableBounds = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any
+): OpenSearchDashboardsDatatable => {
+  const augmentedTable = cloneDeep(datatable);
+  const intervalMillis = moment.duration(dimensions.x.params.interval).asMilliseconds();
+  const xAxisId = getXAxisId(dimensions, augmentedTable.columns);
+  const chartStartTime = new Date(dimensions.x.params.bounds.min).valueOf();
+  const chartEndTime = new Date(dimensions.x.params.bounds.max).valueOf();
+
+  if (!isEmpty(augmentedTable.rows)) {
+    const dataStartTime = augmentedTable.rows[0][xAxisId] as number;
+    const dataEndTime = augmentedTable.rows[augmentedTable.rows.length - 1][xAxisId] as number;
+
+    let curStartTime = dataStartTime;
+    while (curStartTime > chartStartTime) {
+      curStartTime -= intervalMillis;
+      augmentedTable.rows.unshift({
+        [xAxisId]: curStartTime,
+      });
+    }
+
+    let curEndTime = dataEndTime;
+    while (curEndTime < chartEndTime) {
+      curEndTime += intervalMillis;
+      augmentedTable.rows.push({
+        [xAxisId]: curEndTime,
+      });
+    }
+  } else {
+    // if there's no existing rows, create them all
+    let curTime = chartStartTime;
+    while (curTime <= chartEndTime) {
+      augmentedTable.rows.push({
+        [xAxisId]: curTime,
+      });
+      curTime += intervalMillis;
+    }
+  }
+  return augmentedTable;
+};
+
+/**
+ * Adding events into the correct x-axis key (the time bucket)
+ * based on the table. As of now only results from
+ * PointInTimeEventsVisLayers are supported
+ */
+export const addPointInTimeEventsLayersToTable = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any,
+  visLayers: PointInTimeEventsVisLayer[]
+): OpenSearchDashboardsDatatable => {
+  const augmentedTable = addMissingRowsToTableBounds(datatable, dimensions);
+  const xAxisId = getXAxisId(dimensions, augmentedTable.columns);
+
+  if (!isEmpty(visLayers)) {
```

Comment:
nit: 

```js
if (isEmpty(visLayers)) return augmentedTable

// Rest of the code to augment the table
```

Reply:
This would mean adding multiple `return` statements instead of the one. I view the `if` block here as just doing some additional augmentation to the table before the single return. Lmk what you think

Reply:
Thats fine, this suggestion was mostly for readability and is **highly opinionated**. Its easier for me to follow conditional logic that doesnt make me hunt for the closing bracket, especially if im going to be nesting a few if statements. This just helps me reason the code better in my head.

Reply:
Cool, from that context that makes sense to me, i'll add that line for readability

Reply:
Updated in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3145/commits/9c2e3496a8dbfbcbb41a4ae27671e5c0f9c4603f